internal/server/turn.go

```diff
@@ -25,7 +25,7 @@ func handleAllocateRequest(r Request, m *stun.Message) error {
 	//    mechanism of [https://tools.ietf.org/html/rfc5389#section-10.2.2]
 	//    unless the client and server agree to use another mechanism through
 	//    some procedure outside the scope of this document.
-	messageIntegrity, hasAuth, err := authenticateRequest(r, m, stun.MethodAllocate)
+	messageIntegrity, username, hasAuth, err := authenticateRequest(r, m, stun.MethodAllocate)
```

Comment:
Minor stylistic comment: we call `authenticateRequest` from 4 message handlers (allocation request, permission request, channel binding, refresh), but only the first handler needs the username. So I believe it is less intrusive to simply re-parse the username/realm in the allocation handler instead of dropping what's returned at the other 3 invocation sites (see an example in the lifecycle API PR [here](https://github.com/pion/turn/pull/419/files#diff-07ad04f9b305b69b46da66df5d99e7f712864ba650111416305a3fe577f35db1)).

Reply:
Are you sure you'd want to re-iterate over attribute list (linear scan) for the reparse you are suggesting just for the stylistic goal? STUN is small (and sent infrequently) so I don't feel it strongly, but I think it make sense to reuse the artifacts the authorization has just been made with.

Reply:
I'm fine either way, but if we go with your original approach then can you please make it so that we return both the username and the realm from `authenticateRequest` ? The [lifecycle callbacks](https://github.com/pion/turn/pull/419/) that are to be called later during processing the allocate request will need both and it would be an overkill to reparse it (currently we reparse both).

Reply:
In the rebased revision (5cec847):
* Defined authenticationResult to bundle message integrity, username, realm and hasAuth returned by `authenticateRequest`
* Added unit tests for `authenticateRequest`

Reply:
Perfect!